[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/eldanc/mlbootcamp2023/blob/main/lab_3_1_neuralnets.ipynb)

# UofT DSI-CARTE Postdoc Bootcamp
#### Tuesday July 18, 2023
#### Intro to Neural Networks in PyTorch - Lab 1, Day 1
#### Teaching team: Alex Olson, Nakul Upadhya, Shehnaz Islam
##### Lab author: Kyle E. C. Booth, kbooth@mie.utoronto.ca, edited by Jake Mosseri

In this lab, we will be taking our first look at developing our own *neural networks* (NN) with [PyTorch](https://pytorch.org/), probably the most popular machine learning library for working with NNs.

In [14]:
# Install necessary libraries in one command
!pip install torch torchvision torchaudio numpy pandas matplotlib scikit-learn tensorboard
# Import necessary libraries
import numpy as np
import pandas as pd

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.6/5.6 MB 24.6 MB/s eta 0:00:0000:0100:01
  Using cached absl_py-1.4.0-py3-none-any.whl (126 kB)
  Using cached grpcio-1.56.0-cp311-cp311-macosx_10_10_universal2.whl (8.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 181.8/181.8 kB 5.9 MB/s eta 0:00:00
  Using cached google_auth_oauthlib-1.0.0-py2.py3-none-any.whl (18 kB)
  Using cached Markdown-3.4.3-py3-none-any.whl (93 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 400.3/400.3 kB 12.2 MB/s eta 0:00:00
  Using cached tensorboard_data_server-0.7.1-py3-none-macosx_10_9_x86_64.whl (4.8 MB)
  Using cached Werkzeug-2.3.6-py3-none-any.whl (242 kB)
  Using cached cachetools-5.3.1-py3-none-any.whl (9.3 kB)
  Using cached pyasn1_modules-0.3.0-py2.py3-none-any.whl (181 kB)
  Using cached rsa-4.9-py3-none-any.whl (34 kB)
  Using cached requests_oauthlib-1.3.1-py2.py3-none-any.whl (23 kB)
  Using cached pyasn1-0.5.0-py2.py3-none-any.whl (83 kB)
  Using cached oauthlib-3.2.2-py3-none-any.whl (151

### An Intuitive Intro to Neural Nets

In today's lab, we're going to start with an intuitive exercise on the Titanic dataset using Logistic Regression and a simple Neural Network before moving onto some more complex stuff. Let's start by loading our dataset.

The Titanic data is stored in a CSV file (located in the 'data' directory of your root folder), so we need to use Pandas to load the data and then separate it into our X (features) and y (target). We also need to: i) drop unimportant columns, and ii) impute missing values.

We're going to do this all in the next cell.

In [2]:
from sklearn import preprocessing
from sklearn.datasets import fetch_openml

# Fetch the Titanic dataset as a pandas DataFrame
titanic_data = fetch_openml("titanic", version=1, as_frame=True, parser='auto').frame

# Convert 'survived' column to numeric values
titanic_data['survived'] = pd.to_numeric(titanic_data['survived'])

# Remove unnecessary columns
columns_to_drop = ['boat', 'body', 'home.dest']
titanic_data.drop(columns_to_drop, axis=1, inplace=True)

# Remove more unimportant columns
additional_columns_to_drop = ['name', 'ticket', 'cabin', 'embarked']
titanic_data = titanic_data.drop(additional_columns_to_drop, axis=1)

# Encode the 'sex' column using LabelEncoder
label_encoder = preprocessing.LabelEncoder()
label_encoder.fit(titanic_data['sex'])
titanic_data['sex'] = label_encoder.transform(titanic_data['sex'])

# Display the first few rows of the processed data
titanic_data.head()

pclass  survived  sex      age  sibsp  parch      fare
0       1         1    0  29.0000      0      0  211.3375
1       1         1    1   0.9167      1      2  151.5500
2       1         0    0   2.0000      1      2  151.5500
3       1         0    1  30.0000      1      2  151.5500
4       1         0    0  25.0000      1      2  151.5500

Next, we will split the dataset into a training set (where we will do our cross validation) and a test set (our hold-out data).

In [3]:
from sklearn.model_selection import train_test_split

# Split dataset into features and target
target_data = titanic_data["survived"]
feature_data = titanic_data.drop("survived", axis=1)

# Split data into training set and test set
X_train, X_test, y_train, y_test = train_test_split(feature_data, target_data, test_size=0.3, random_state=0)

Now, we're ready to try out some models on our training data. Since we're solving a binary classification problem (i.e., predicting a 0 or 1 target), we want to design classifiers. Examples of simple machine learning classifiers include: **k-nearest neighbors**, **decision trees**, and **logistic regression**.

In this exercise, we're going to fit a logistic regression to our data and then design a neural network architecture that behaves exactly like a logistic regression and validate that we get the same result.

##### Intro: Logistic Regression

Logistic regression models are linear models similar to linear regression models. The linear regression equation can be written as follows:


<center>$\hat{y} = \beta_0 + \beta_1 x_1 + \beta_2 x_2 +, \dots, + \beta_n x_n$</center>

Where $\hat{y}$ is our prediction, $\beta$ is our vector of coefficients (the things we learn), and $x$ is our feature vector. The linear regression equation defines a line in $n$ dimensional space. The problem with linear regression is that it doesn't really perform well on classification tasks. Consider the following example:

<img src="https://github.com/lyeskhalil/mlbootcamp/blob/master/img/linear-classification.png?raw=1" width="500"/>

The green line represents our trained linear regression model. Our feature is the size of a tumor, and our target is whether it is malignant or not (0 or 1). As we can see, even though our model is trained to the data to minimize error, for a lot of the values of tumor size it is going to give us a weird result (e.g., for some really small tumors, the prediction would be a negative value!).

To resolve this, we use the *logistic function* (also called the *sigmoid* function) to 'squish' our linear model to be bounded by 0 and 1. The logistic (sigmoid) function is $\frac{1}{1+e^{-x}}$, and thus our logistic regression equation becomes:

<center>$\large\hat{y} = \frac{1}{1+e^{-(\beta_0 + \beta_1 x_1 + \beta_2 x_2 +, \dots, + \beta_n x_n)}}$</center>

In this equation, the values for $\hat{y}$ can never be below 0, and can never exceed 1 even for the most extreme feature values $x$.

**YOUR TURN:**
* Assuming you've trained a nice logistic regression model to the below data (see Figure), what might the model fit look like (i.e., what will the line look like)? ____________________________________
* For new data samples with features $x$, how would you convert the output of the logistic regression, $\hat{y}$, into a classification (0 or 1)? ______________________________

<img src="https://github.com/lyeskhalil/mlbootcamp/blob/master/img/logistic-classification.jpg?raw=1" width="400"/>

OK, cool! Let's use scikit-learn to fit a logistic regression model to our training set and then predict on our test set (we won't do cross validation this time).

Before we can fit our logistic regression model to our training data, we can conduct imputation to replace missing values with the mean/median/mode value in the column. For this exercise we will conduct mode imputation (i.e., the most common value in the column).


In [4]:
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Impute missing values using the most frequent strategy
imp = SimpleImputer(missing_values=np.nan, strategy='most_frequent')
imp.fit(X_train)

# Transform both the training and test sets
X_train = imp.transform(X_train)
X_test = imp.transform(X_test)

# Fit the logistic regression model
logreg = LogisticRegression()
logreg.fit(X_train, y_train)

# Make predictions on the test set
predictions = logreg.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, predictions)
print(f"Logistic accuracy: {accuracy*100:.2f}%")

Logistic accuracy: 79.13%


The results show our logistic regression is 79.13% accurate in its predictions.

**YOUR TURN:**
* The default regularization parameter for sklearn's logistic regression is L2 (or ridge regression); can you figure out how to change it to L1 (LASSO)? _Note: you'll have to look at the [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html)!_ ______________________
* What is the mean accuracy of an L1 regularized Logistic regression model on the training set? ______________________

OK - time for the good stuff!

### Neural Networks

A *neural network* (NN) is a type of machine learning model that, like linear or logistic regression, takes a feature vector, X, as input and predicts a target, y. The way it does this is a little bit different, however. A typical NN architecture consists of: an input layer, hidden layers, and an output layer. Each layer consists of a set of nodes (neurons) connected by edges (outputs). Let's look at the figure below:

<img src="https://github.com/lyeskhalil/mlbootcamp/blob/master/img/nn.jpeg?raw=1" width="400"/>

**Input layer**: This is a passive layer that simply takes in your feature data and outputs it to the hidden layers. You can think of each input layer neuron as being associated with a feature in your feature set.

**Hidden layer**: This is where the magic happens. The original features, as received by the input layer, go through a series of transformations within the hidden layer. You can think of each node (neuron) within the hidden layer as a highly transformed feature.

**Output layer**: This is where we get our final result, the 0 or 1 prediction.

### Zooming In

Let's take a look at what is happening at any given node (neuron) within the hidden layer. Take a look at the following image of a neuron within an NN:

<img src="https://github.com/lyeskhalil/mlbootcamp/blob/master/img/neuron.png?raw=1" width="500"/>

Every neuron has some inputs ($x_1, x_2, \dots, x_n$) with input weights ($w_1, w_2, \dots, w_n$) and an output, $Y$. The neuron itself applies a transformation, $f$, known as the *activation function*, to the linear combination of its inputs and input weights. The value $b$ is a constant weight called the bias.

There are many different types of activation functions, but the popular ones are the *sigmoid*, *tanh*, and *ReLU* activation functions. Yes, you heard correctly: the sigmoid function is a popular activation function! (This should be reminding you of the logistic regression model we discussed above).

**YOUR TURN:**
* If you were to develop a simple neural network architecture that was equivalent to a logistic regression model for the Titanic data, how would you do it? Get a pen and paper and draw it out. Make sure to specify: the input layer, the hidden layer(s), the output layer, the activation function(s), the weights, and the biases.
* How many hidden layers does your NN have? What type of activation function, $f$, does it use? _________________
* Say you wanted to add another layer to your NN architecture with 3x neurons, what would your new architecture look like? _________________

### Intro to  PyTorch

OK, so now that we've made the connection between NNs and Logistic Regression, let's code up our little NN in PyTorch and use it to predict survivorship on the Titanic dataset.

First, *tensors* are the fundamental data type of PyTorch. Each tensor is effectively a multi-dimensional array, just like a numpy array. The primary difference is that tensors have been setup in such a way to enhance the NN training process.

Let's load our X and y training data into tensors:

In [5]:
import torch

X_train_tensor = torch.tensor(X_train).float()
y_train_tensor = torch.tensor(y_train.values).float()

X_test_tensor = torch.tensor(X_test).float()
y_test_tensor = torch.tensor(y_test.values).float()

In [7]:
print("Shape of X_train_tensor:", X_train_tensor.shape)
print("Shape of y_train_tensor:", y_train_tensor.shape)
print("Shape of X_test_tensor:", X_test_tensor.shape)
print("Shape of y_test_tensor:", y_test_tensor.shape)

Shape of X_train_tensor: torch.Size([916, 6])
Shape of y_train_tensor: torch.Size([916])
Shape of X_test_tensor: torch.Size([393, 6])
Shape of y_test_tensor: torch.Size([393])


Next, we will actually define our logistic regression network model class. The below function, `LogisticRegression`, applies a sigmoid transformation to the output, as required.

In [8]:
import torch

class LogisticRegression(torch.nn.Module):
    def __init__(self, input_dim, output_dim):
        super(LogisticRegression, self).__init__()
        self.linear = torch.nn.Linear(input_dim, output_dim)

    def forward(self, x):
        outputs = torch.sigmoid(self.linear(x))
        return outputs

Next, we identify the dimensions of our problem: 6 x 1 (6 features and 2 target classes: 0 or 1), initialize our model with those dimensions and then specify the loss function ([cross entropy](https://en.wikipedia.org/wiki/Cross_entropy)) and optimization technique ([stochastic gradient descent](https://en.wikipedia.org/wiki/Stochastic_gradient_descent)).

In [9]:
input_dim = 6
output_dim = 1

model = LogisticRegression(input_dim, output_dim)

criterion = torch.nn.BCELoss()
optimizer = torch.optim.SGD(model.parameters(), lr=0.001)

In [10]:
model

LogisticRegression(
  (linear): Linear(in_features=6, out_features=1, bias=True)
)

Next, we set up our dataset to be *iterable* such that we can train our neural network in *batches*. A batch is a subset of the total data such that if we combined them all, we'd get the whole dataset. Batching is done to speed up the training process and reduce memory requirements.

**YOUR TURN:**
* If we select a batch size of 256, how many batches of training data will be generated?________________

The `DataLoader` function does this batching operation for us.

In [11]:
batch_size = 256

train_dataset = torch.utils.data.TensorDataset(X_train_tensor, y_train_tensor)
train_loader = torch.utils.data.DataLoader(dataset=train_dataset, batch_size=batch_size, shuffle=True)

Next, we train our NN and test its performance on the test set every 4000 iterations. We use 40,000 *epochs* and print our accuracy values every 4000 iterations. An epoch is when the entire dataset has passed through the network. An iteration is when a single forward/backward pass of the network over a batch of data is done.

**YOUR TURN:**
* If our batch size is 256 and we elect to do 40,000 epochs (i.e., the network sees the entire dataset 40,000 times), how many iterations (forward/backward passes of the data) will our network see? _______________________

Run the below code and check out the incremental accuracy improvement output below.

In [15]:
import random
from torch.utils.tensorboard import SummaryWriter

random.seed(1)
torch.manual_seed(1)
torch.backends.cudnn.deterministic = True

iterations = 0
epochs = 40000

# Create a SummaryWriter to write TensorBoard logs
writer = SummaryWriter()

for epoch in range(epochs):
    for batch_idx, (features, target) in enumerate(train_loader):
        optimizer.zero_grad()
        outputs = model(features)
        loss = criterion(torch.squeeze(outputs), target)
        loss.backward()
        optimizer.step()

        iterations += 1
        if iterations % 4000 == 0:
            model.eval()
            with torch.no_grad():
                outputs = model(X_test_tensor)
                predicted = torch.round(outputs.data).numpy()
            model.train()

            accuracy = accuracy_score(y_test, predicted)
            progress = epoch / epochs * 100

            # Write loss and accuracy to TensorBoard
            writer.add_scalar("Loss", loss.item(), iterations)
            writer.add_scalar("Accuracy", accuracy, iterations)

            print(f"Loss: {loss.item():.2f} | Accuracy: {accuracy:.2f} | Progress: {progress:.2f}% | Iteration: {iterations}")

# Close the SummaryWriter
writer.close()

Loss: 0.47 | Accuracy: 0.78 | Progress: 2.50% | Iteration: 4000
Loss: 0.43 | Accuracy: 0.78 | Progress: 5.00% | Iteration: 8000
Loss: 0.51 | Accuracy: 0.78 | Progress: 7.50% | Iteration: 12000
Loss: 0.43 | Accuracy: 0.78 | Progress: 10.00% | Iteration: 16000
Loss: 0.48 | Accuracy: 0.78 | Progress: 12.50% | Iteration: 20000
Loss: 0.46 | Accuracy: 0.78 | Progress: 15.00% | Iteration: 24000
Loss: 0.49 | Accuracy: 0.78 | Progress: 17.50% | Iteration: 28000
Loss: 0.37 | Accuracy: 0.78 | Progress: 20.00% | Iteration: 32000
Loss: 0.50 | Accuracy: 0.78 | Progress: 22.50% | Iteration: 36000
Loss: 0.52 | Accuracy: 0.78 | Progress: 25.00% | Iteration: 40000
Loss: 0.43 | Accuracy: 0.78 | Progress: 27.50% | Iteration: 44000
Loss: 0.50 | Accuracy: 0.78 | Progress: 30.00% | Iteration: 48000
Loss: 0.43 | Accuracy: 0.79 | Progress: 32.50% | Iteration: 52000
Loss: 0.50 | Accuracy: 0.78 | Progress: 35.00% | Iteration: 56000
Loss: 0.48 | Accuracy: 0.78 | Progress: 37.50% | Iteration: 60000
Loss: 0.55 | Ac

KeyboardInterrupt: 

In [16]:
# Load the TensorBoard notebook extension
%load_ext tensorboard

Hopefully you got an accuracy of around ~78%. What you'll notice is that is similar accuracy we got from sklearn's built-in logistic regression function from earlier in the lab. There are differences in the implementation that account for this discrepancy, notably sklearn used l2 regularization, and an optimizer called LBFGS instead of SGD that we learned and used.

Let's take a look at the trained model parameters using the `model.parameters()` function within PyTorch.

In [17]:
params = list(model.parameters())
params

[Parameter containing:
 tensor([[-0.4100, -2.3163, -0.0100, -0.3420, -0.0245,  0.0062]],
        requires_grad=True),
 Parameter containing:
 tensor([1.9463], requires_grad=True)]

Hm, well this is interesting! We can see that our model consists of two tensors: the first has (1,6), and the second has dimension (1). Refer back to how you drew what you thought this NN architecture would look like.

**YOU TURN:**
* What do you think these values represent? ____________________________
* How many hidden layers does the architecture have? ______________________________
* Draw the architecture and label (some of) the weights (trained parameters). ______________________________

(*Hint: to answer these questions, try printing `outputs.data[0]` and `predicted[0]` to look at the model's assessment of the first sample*)

You'll notice that this took considerably longer to train than scikit-learn's logistic regression: this is because PyTorch is set-up to be more flexible and train architectures much more complex than a simple single neuron network. Scikit-learn's implementation of logistic regression is highly optimized.

**YOUR TURN:**
* How many epochs would you need to increase the process to 200,000 iterations? ______________________
* Does increasing to 200,000 iterations improve your test set accuracy? ______________________
* Compare the predictions of your logistic regression from scikit-learn and your network developed with PyTorch. Are all the predictions the same? How many predictions are pair-wise different? ______________________

* Create a simple three layer fully connected neural network with a single hidden layer of 4 neurons followed by Relu activation, and learning rate of 0.01. What is the final accuracy of this network? ______________________
<br>(Hint: If you are using **BCELoss** as loss criterion, you need to use a **sigmoid** activation function for the output layer. If you are using **CrossEntropyLoss** as loss criterion, you need to use a **softmax** activation function for the output layer. While BCELoss is used for binary classification, CrossEntropyLoss is used for multi-class classification.)

Congratulations! You've completed an introduction to neural networks and PyTorch. If you want to explore more sophisicated architectures and applications, check out designing a PyTorch neural network to properly classify digit images here:

https://towardsdatascience.com/handwritten-digit-mnist-pytorch-977b5338e627

Other than that, you're done the lab!